File: lexer/src/lexer.rs
use std::{
    cell::RefCell,
    iter::Peekable,
    str::Chars,
};

use diag::{
    Diagnostic,
    Label,
    Severity,
};
use session::{
    Session,
    interner::syms,
};
use span::SourceSpan;
use unicode_ident::{
    is_xid_continue,
    is_xid_start,
};

use super::Token;
use crate::{
    AssignmentKind,
    LiteralKind,
    TokenKind,
};

#[derive(Debug)]
pub enum LexerError {
    UnknownToken(String),
    UnknownEscapeString,
    UnclosedString,
}

// NOTE: the lexer error display isn't used by anything.
impl std::fmt::Display for LexerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{self:?}")
    }
}

impl std::error::Error for LexerError {}

pub struct Lexer<'sess> {
    session: &'sess Session,
    current: Option<char>,
    chars: Peekable<Chars<'sess>>,

    /// The current byte offset the lexer is at.
    current_offset: usize,
}

impl<'sess> Lexer<'sess> {
    pub fn new(session: &'sess Session) -> Lexer<'sess> {
        let mut chars = session.source_text.chars().peekable();
        let current = chars.next();

        Lexer {
            session,
            current,
            chars,
            current_offset: 0,
        }
    }

    fn advance(&mut self) -> Option<char> {
        let result = self.current;
        if let Some(c) = result {
            self.current_offset += c.len_utf8();
        }
        self.current = self.chars.next();
        result
    }

    pub fn next_token(&mut self) -> Result<Token, LexerError> {
        loop {
            match self.current {
                // skips all lines that start with `#`
                Some('#') => {
                    while let Some(c) = self.current {
                        if c == '\n' {
                            break;
                        }
                        self.advance();
                    }
                },
                // skips all empty whitespaces
                Some(' ' | '\t' | '\r') => {
                    self.advance();
                },
                // skips newlines
                Some('\n') => {
                    self.advance();
                },
                // break the loop if it isn't a whitespace or a comment
                _ => break,
            };
        }

        let start_offset = self.current_offset;

        if self.current.is_none() {
            return Ok(Token {
                span: SourceSpan::new(start_offset, start_offset),
                kind: TokenKind::EOF,
            });
        }

        let mut token = match self.current {
            Some(':') => {
                self.advance();
                Ok(Token {
                    span: SourceSpan::default(),
                    kind: TokenKind::Colon,
                })
            },
            Some('=') => {
                self.advance();
                match self.current {
                    Some('=') => {
                        self.advance();
                        Ok(Token {
                            span: SourceSpan::default(),
                            kind: TokenKind::Eq,
                        })
                    },
                    _ => Ok(Token {
                        span: SourceSpan::default(),
                        kind: TokenKind::Assign {
                            kind: AssignmentKind::Assign,
                        },
                    }),
                }
            },
            Some('!') => {
                self.advance();
                match self.current {
                    Some('=') => {
                        self.advance();
                        Ok(Token {
                            span: SourceSpan::default(),
                            kind: TokenKind::Ne,
                        })
                    },
                    _ => Ok(Token {
                        span: SourceSpan::default(),
                        kind: TokenKind::Not,
                    }),
                }
            },
            Some('&') => {
                self.advance();
                match self.current {
                    Some('&') => {
                        self.advance();
                        Ok(Token {
                            span: SourceSpan::default(),
                            kind: TokenKind::And,
                        })
                    },
                    Some('=') => {
                        self.advance();
                        Ok(Token {
                            span: SourceSpan::default(),
                            kind: TokenKind::Assign {
                                kind: AssignmentKind::BitAndAssign,
                            },
                        })
                    },
                    _ => Ok(Token {
                        span: SourceSpan::default(),
                        kind: TokenKind::BitAnd,
                    }),
                }
            },
            Some('|') => {
                self.advance();
                match self.current {
                    Some('|') => {
                        self.advance();
                        Ok(Token {
                            span: SourceSpan::default(),
                            kind: TokenKind::Or,
                        })
                    },
                    Some('=') => {
                        self.advance();
                        Ok(Token {
                            span: SourceSpan::default(),
                            kind: TokenKind::Assign {
                                kind: AssignmentKind::BitOrAssign,
                            },
                        })
                    },
                    _ => Ok(Token {
                        span: SourceSpan::default(),
                        kind: TokenKind::BitOr,
                    }),
                }
            },
            Some('^') => {
                self.advance();
                match self.current {
                    Some('=') => {
                        self.advance();
                        Ok(Token {
                            span: SourceSpan::default(),
                            kind: TokenKind::Assign {
                                kind: AssignmentKind::BitXorAssign,
                            },
                        })
                    },
                    _ => Ok(Token {
                        span: SourceSpan::default(),
                        kind: TokenKind::BitXor,
                    }),
                }
            },
            Some('<') => {
                self.advance();
                match self.current {
                    Some('=') => {
                        self.advance();
                        Ok(Token {
                            span: SourceSpan::default(),
                            kind: TokenKind::Le,
                        })
                    },
                    Some('<') => {
                        self.advance();
                        match self.chars.peek() {
                            Some('=') => {
                                self.advance();
                                Ok(Token {
                                    span: SourceSpan::default(),
                                    kind: TokenKind::Assign {
                                        kind: AssignmentKind::ShiftLeftAssign,
                                    },
                                })
                            },
                            _ => Ok(Token {
                                span: SourceSpan::default(),
                                kind: TokenKind::ShiftLeft,
                            }),
                        }
                    },
                    _ => Ok(Token {
                        span: SourceSpan::default(),
                        kind: TokenKind::Lt,
                    }),
                }
            },
            Some('>') => {
                self.advance();
                match self.current {
                    Some('=') => {
                        self.advance();
                        Ok(Token {
                            span: SourceSpan::default(),
                            kind: TokenKind::Ge,
                        })
                    },
                    Some('>') => {
                        self.advance();
                        match self.chars.peek() {
                            Some('=') => {
                                self.advance();
                                Ok(Token {
                                    span: SourceSpan::default(),
                                    kind: TokenKind::Assign {
                                        kind: AssignmentKind::ShiftRightAssign,
                                    },
                                })
                            },
                            _ => Ok(Token {
                                span: SourceSpan::default(),
                                kind: TokenKind::ShiftLeft,
                            }),
                        }
                    },
                    _ => Ok(Token {
                        span: SourceSpan::default(),
                        kind: TokenKind::Gt,
                    }),
                }
            },
            Some('+') => {
                self.advance();
                match self.current {
                    Some('=') => {
                        self.advance();
                        Ok(Token {
                            span: SourceSpan::default(),
                            kind: TokenKind::Assign {
                                kind: AssignmentKind::AddAssign,
                            },
                        })
                    },
                    _ => Ok(Token {
                        span: SourceSpan::default(),
                        kind: TokenKind::Add,
                    }),
                }
            },
            Some('-') => {
                self.advance();
                match self.current {
                    Some('=') => {
                        self.advance();
                        Ok(Token {
                            span: SourceSpan::default(),
                            kind: TokenKind::Assign {
                                kind: AssignmentKind::SubAssign,
                            },
                        })
                    },
                    _ => Ok(Token {
                        span: SourceSpan::default(),
                        kind: TokenKind::Sub,
                    }),
                }
            },
            Some('*') => {
                self.advance();
                match self.current {
                    Some('=') => {
                        self.advance();
                        Ok(Token {
                            span: SourceSpan::default(),
                            kind: TokenKind::Assign {
                                kind: AssignmentKind::MulAssign,
                            },
                        })
                    },
                    _ => Ok(Token {
                        span: SourceSpan::default(),
                        kind: TokenKind::Mul,
                    }),
                }
            },
            Some('/') => {
                self.advance();
                match self.current {
                    Some('=') => {
                        self.advance();
                        Ok(Token {
                            span: SourceSpan::default(),
                            kind: TokenKind::Assign {
                                kind: AssignmentKind::DivAssign,
                            },
                        })
                    },
                    _ => Ok(Token {
                        span: SourceSpan::default(),
                        kind: TokenKind::Div,
                    }),
                }
            },
            Some('%') => {
                self.advance();
                match self.current {
                    Some('=') => {
                        self.advance();
                        Ok(Token {
                            span: SourceSpan::default(),
                            kind: TokenKind::Assign {
                                kind: AssignmentKind::ModAssign,
                            },
                        })
                    },
                    _ => Ok(Token {
                        span: SourceSpan::default(),
                        kind: TokenKind::Mod,
                    }),
                }
            },
            Some('(') => {
                self.advance();
                Ok(Token {
                    span: SourceSpan::default(),
                    kind: TokenKind::LeftParen,
                })
            },
            Some(')') => {
                self.advance();
                Ok(Token {
                    span: SourceSpan::default(),
                    kind: TokenKind::RightParen,
                })
            },
            Some('{') => {
                self.advance();
                Ok(Token {
                    span: SourceSpan::default(),
                    kind: TokenKind::LeftBrace,
                })
            },
            Some('}') => {
                self.advance();
                Ok(Token {
                    span: SourceSpan::default(),
                    kind: TokenKind::RightBrace,
                })
            },
            Some('[') => {
                self.advance();
                Ok(Token {
                    span: SourceSpan::default(),
                    kind: TokenKind::LeftBracket,
                })
            },
            Some(']') => {
                self.advance();
                Ok(Token {
                    span: SourceSpan::default(),
                    kind: TokenKind::RightBracket,
                })
            },
            Some(';') => {
                self.advance();
                Ok(Token {
                    span: SourceSpan::default(),
                    kind: TokenKind::Semicolon,
                })
            },
            Some(',') => {
                self.advance();
                Ok(Token {
                    span: SourceSpan::default(),
                    kind: TokenKind::Comma,
                })
            },
            Some('.') => {
                self.advance();
                Ok(Token {
                    span: SourceSpan::default(),
                    kind: TokenKind::Dot,
                })
            },
            Some('\\') => {
                self.advance();
                Ok(Token {
                    span: SourceSpan::default(),
                    kind: TokenKind::Backslash,
                })
            },
            Some('"') => self.read_string(),
            Some(c) if c.is_numeric() => self.read_number(),
            Some(_) => self.read_identifier(),
            _ => unreachable!(),
        }?;

        token.span = SourceSpan::new(start_offset, self.current_offset);
        Ok(token)
    }

    fn read_string(&mut self) -> Result<Token, LexerError> {
        let string_start = self.current_offset;
        self.advance(); // consume the opening "
        let mut result = String::new();

        loop {
            match self.advance() {
                Some('\\') => {
                    let escape_start = self.current_offset - 1;
                    match self.current {
                        Some('n') => {
                            self.advance();
                            result.push('\n');
                        },
                        Some('r') => {
                            self.advance();
                            result.push('\r');
                        },
                        Some('t') => {
                            self.advance();
                            result.push('\t');
                        },
                        Some('"') => {
                            self.advance();
                            result.push('"');
                        },
                        Some('\\') => {
                            self.advance();
                            result.push('\\');
                        },
                        Some('x') => {
                            self.advance(); // consume the 'x'

                            let hi = self.advance().and_then(|c| c.to_digit(16)).map(|d| d as u8);
                            let lo = self.advance().and_then(|c| c.to_digit(16)).map(|d| d as u8);

                            match (hi, lo) {
                                (Some(hi), Some(lo)) => result.push(((hi << 4) | lo) as char),
                                (_, _) => {
                                    let span = SourceSpan::new(escape_start, self.current_offset);
                                    self.session.emit(
                                        Diagnostic::error("Unknown escape string")
                                            .with_label(Label::primary(span, "Unknown escape string")),
                                    );
                                    return Err(LexerError::UnknownEscapeString);
                                },
                            }
                        },
                        Some(_) => {
                            self.advance();
                            let span = SourceSpan::new(escape_start, self.current_offset);
                            self.session.emit(
                                Diagnostic::error("Unknown escape string")
                                    .with_label(Label::primary(span, "Unknown escape string")),
                            );
                            return Err(LexerError::UnknownEscapeString);
                        },
                        None => {
                            let span = SourceSpan::new(string_start, self.current_offset);
                            self.session.emit(
                                Diagnostic::error("Unclosed string")
                                    .with_label(Label::primary(span, "Unclosed string")),
                            );
                            return Err(LexerError::UnclosedString);
                        },
                    }
                },
                Some('"') => break,
                Some(c) => result.push(c),
                None => {
                    let span = SourceSpan::new(string_start, self.current_offset);
                    self.session
                        .emit(Diagnostic::error("Unclosed string").with_label(Label::primary(span, "Unclosed string")));
                    return Err(LexerError::UnclosedString);
                },
            }
        }

        let sym = self.session.intern_string(&result);

        Ok(Token {
            span: SourceSpan::default(),
            kind: TokenKind::Literal {
                kind: LiteralKind::String,
                sym,
            },
        })
    }

    fn read_identifier(&mut self) -> Result<Token, LexerError> {
        let start_offset = self.current_offset;
        match self.current {
            Some(c) if is_xid_start(c) => {
                self.advance();

                while let Some(c) = self.current {
                    if is_xid_continue(c) {
                        self.advance();
                    } else {
                        break;
                    }
                }

                let identifier = &self.session.source_text[start_offset..self.current_offset];
                let sym = self.session.intern_string(&identifier);
                let kind = match sym {
                    syms::FN => TokenKind::Function,
                    syms::WHILE => TokenKind::While,
                    syms::BREAK => TokenKind::KwBreak,
                    syms::CONTINUE => TokenKind::KwContinue,
                    syms::TRUE => TokenKind::KwTrue,
                    syms::FALSE => TokenKind::KwFalse,
                    syms::IF => TokenKind::If,
                    syms::ELSE => TokenKind::Else,
                    syms::RETURN => TokenKind::Return,
                    syms::STRUCT => TokenKind::Struct,
                    _ => TokenKind::Ident { sym },
                };

                Ok(Token {
                    kind,
                    span: SourceSpan::default(),
                })
            },
            Some(c) => {
                let char_len = c.len_utf8();
                let span = SourceSpan::new(self.current_offset, self.current_offset + char_len);
                self.session
                    .emit(Diagnostic::error("Unknown token").with_label(Label::primary(span, "Unknown token")));
                Err(LexerError::UnknownToken(c.to_string()))
            },
            _ => Ok(Token {
                span: SourceSpan::default(),
                kind: TokenKind::EOF,
            }),
        }
    }

    fn read_number(&mut self) -> Result<Token, LexerError> {
        let mut has_decimal = false;
        let mut number = String::new();

        while let Some(c) = self.current {
            if c.is_ascii_digit() {
                number.push(c);
                self.advance();
            } else if c == '.' && !has_decimal {
                has_decimal = true;
                number.push(c);
                self.advance();
            } else {
                break;
            }
        }

        let kind = if has_decimal {
            LiteralKind::Float
        } else {
            LiteralKind::Integer
        };

        let sym = self.session.intern_string(&number);

        Ok(Token {
            span: SourceSpan::default(),
            kind: TokenKind::Literal { kind, sym },
        })
    }
}

File: lib/Lexer/Lexer.cpp
#include "belalang/Lexer/Lexer.h"
#include "llvm/Support/ConvertUTF.h"

namespace belalang {
namespace lexer {

static char32_t decodeUTF8(const char *&ptr, const char *end) {
  if (ptr >= end)
    return 0;

  const auto *sourcePtr = reinterpret_cast<const llvm::UTF8 *>(ptr);
  const auto *sourceEnd = reinterpret_cast<const llvm::UTF8 *>(end);
  llvm::UTF32 codePoint = 0;

  llvm::ConversionResult res = llvm::convertUTF8Sequence(
      &sourcePtr, sourceEnd, &codePoint, llvm::strictConversion);

  if (res == llvm::conversionOK) {
    ptr = reinterpret_cast<const char *>(sourcePtr);
    return static_cast<char32_t>(codePoint);
  }

  auto fallback = static_cast<unsigned char>(*ptr);
  ++ptr;
  return fallback;
}

bool Lexer::lex(Token &result) {
  result.startToken();

  while (ptr < end) {
    if (*ptr == ' ' || *ptr == '\t' || *ptr == '\r' || *ptr == '\n') {
      ++ptr;
    } else if (*ptr == '#') {
      while (ptr < end && *ptr != '\n') {
        ++ptr;
      }
    } else {
      break;
    }
  }

  if (ptr >= end) {
    result.setKind(TokenKind::EndOfFile);
    return false;
  }

  offsetStart = ptr - start;
  const char *prevPtr = ptr;
  char32_t cp = decodeUTF8(ptr, end);

  switch (cp) {
  case ':':
    result.setKind(TokenKind::Colon);
    break;

  case '=':
    if (ptr < end) {
      if (*ptr == '=') {
        ++ptr;
        result.setKind(TokenKind::Eq);
        break;
      }
    }
    result.setKind(TokenKind::Assign);
    break;

  case '!':
    if (ptr < end) {
      if (*ptr == '=') {
        ++ptr;
        result.setKind(TokenKind::Ne);
        break;
      }
    }
    result.setKind(TokenKind::Not);
    break;

  case '&':
    if (ptr < end) {
      if (*ptr == '&') {
        ++ptr;
        result.setKind(TokenKind::And);
        break;
      }
      if (*ptr == '=') {
        ++ptr;
        result.setKind(TokenKind::BitAndAssign);
        break;
      }
    }
    result.setKind(TokenKind::BitAnd);
    break;

  case '|':
    if (ptr < end) {
      if (*ptr == '|') {
        ++ptr;
        result.setKind(TokenKind::Or);
        break;
      }
      if (*ptr == '=') {
        ++ptr;
        result.setKind(TokenKind::BitOrAssign);
        break;
      }
    }
    result.setKind(TokenKind::BitOr);
    break;

  case '^':
    if (ptr < end) {
      if (*ptr == '=') {
        ++ptr;
        result.setKind(TokenKind::BitXorAssign);
        break;
      }
    }
    result.setKind(TokenKind::BitXor);
    break;

  case '<':
    if (ptr < end) {
      if (*ptr == '=') {
        ++ptr;
        result.setKind(TokenKind::Le);
        break;
      }
      if (*ptr == '<') {
        ++ptr;
        if (ptr < end) {
          if (*ptr == '=') {
            ++ptr;
            result.setKind(TokenKind::ShiftLeftAssign);
            break;
          }
        }
        result.setKind(TokenKind::ShiftLeft);
        break;
      }
    }
    result.setKind(TokenKind::Lt);
    break;

  case '>':
    if (ptr < end) {
      if (*ptr == '=') {
        ++ptr;
        result.setKind(TokenKind::Ge);
        break;
      }
      if (*ptr == '>') {
        ++ptr;
        if (ptr < end) {
          if (*ptr == '=') {
            ++ptr;
            result.setKind(TokenKind::ShiftRightAssign);
            break;
          }
        }
        result.setKind(TokenKind::ShiftRight);
        break;
      }
    }
    result.setKind(TokenKind::Gt);
    break;

  case '+':
    if (ptr < end) {
      if (*ptr == '=') {
        ++ptr;
        result.setKind(TokenKind::AddAssign);
        break;
      }
    }
    result.setKind(TokenKind::Add);
    break;

  case '-':
    if (ptr < end) {
      if (*ptr == '=') {
        ++ptr;
        result.setKind(TokenKind::SubAssign);
        break;
      }
    }
    result.setKind(TokenKind::Sub);
    break;

  case '*':
    if (ptr < end) {
      if (*ptr == '=') {
        ++ptr;
        result.setKind(TokenKind::MulAssign);
        break;
      }
    }
    result.setKind(TokenKind::Mul);
    break;

  case '/':
    if (ptr < end) {
      if (*ptr == '=') {
        ++ptr;
        result.setKind(TokenKind::DivAssign);
        break;
      }
    }
    result.setKind(TokenKind::Div);
    break;

  case '%':
    if (ptr < end) {
      if (*ptr == '=') {
        ++ptr;
        result.setKind(TokenKind::ModAssign);
        break;
      }
    }
    result.setKind(TokenKind::Mod);
    break;

  case '(':
    result.setKind(TokenKind::LeftParen);
    break;

  case ')':
    result.setKind(TokenKind::RightParen);
    break;

  case '{':
    result.setKind(TokenKind::LeftBrace);
    break;

  case '}':
    result.setKind(TokenKind::RightBrace);
    break;

  case '[':
    result.setKind(TokenKind::LeftBracket);
    break;

  case ']':
    result.setKind(TokenKind::RightBracket);
    break;

  case ';':
    result.setKind(TokenKind::Semicolon);
    break;

  case ',':
    result.setKind(TokenKind::Comma);
    break;

  case '.':
    result.setKind(TokenKind::Dot);
    break;

  case '\\':
    result.setKind(TokenKind::Backslash);
    break;

  case '"':
    if (!lexString(result))
      return false;
    break;

  default:
    if (isIdentStart(cp)) {
      if (!lexIdentifier(result, prevPtr))
        return false;
      break;
    }
    if (cp >= '0' && cp <= '9') {
      if (!lexNumber(result, static_cast<char>(cp)))
        return false;
      break;
    }
    diag.print(diag::Diagnostic::error("Unknown token")
                   .withLabel(diag::Label::primary(
                       offsetStart, (size_t)(ptr - start), "Unknown token")));
    return false;
  }

  result.setSpan(offsetStart, (size_t)(ptr - start));
  return true;
}

bool Lexer::lexString(Token &result) {
  std::string str;

  while (ptr < end) {
    char c = *ptr;
    if (c == '\\') {
      size_t escapeStart = (size_t)(ptr - start);
      ++ptr;
      if (ptr >= end) {
        diag.print(diag::Diagnostic::error("Unclosed string")
                       .withLabel(diag::Label::primary(offsetStart,
                                                       (size_t)(ptr - start),
                                                       "Unclosed string")));
        return false;
      }

      switch (*ptr) {
      case 'n':
        str += '\n';
        ++ptr;
        break;
      case 'r':
        str += '\r';
        ++ptr;
        break;
      case 't':
        str += '\t';
        ++ptr;
        break;
      case '"':
        str += '"';
        ++ptr;
        break;
      case '\\':
        str += '\\';
        ++ptr;
        break;
      case 'x': {
        ++ptr; // consume the `x`

        if (ptr >= end || !isxdigit(static_cast<unsigned char>(*ptr))) {
          diag.print(diag::Diagnostic::error("Unknown escape string")
                         .withLabel(diag::Label::primary(
                             escapeStart, (size_t)(ptr - start) + 1,
                             "Unknown escape string")));
          return false;
        }
        unsigned char hi = static_cast<unsigned char>(*ptr++);

        if (ptr >= end || !isxdigit(static_cast<unsigned char>(*ptr))) {
          diag.print(diag::Diagnostic::error("Unknown escape string")
                         .withLabel(diag::Label::primary(
                             escapeStart, (size_t)(ptr - start) + 1,
                             "Unknown escape string")));
          return false;
        }
        unsigned char lo = static_cast<unsigned char>(*ptr++);

        auto hexVal = [](unsigned char c) -> unsigned char {
          if (c >= '0' && c <= '9')
            return c - '0';
          if (c >= 'a' && c <= 'f')
            return c - 'a' + 10;
          return c - 'A' + 10;
        };

        str += static_cast<char>((hexVal(hi) << 4) | hexVal(lo));
        break;
      }
      default:
        diag.print(diag::Diagnostic::error("Unknown escape string")
                       .withLabel(diag::Label::primary(
                           escapeStart, (size_t)(ptr - start) + 1,
                           "Unknown escape string")));
        return false;
      }
    } else if (c == '"') {
      ++ptr;
      result.setKind(TokenKind::LiteralString);
      result.setStr(str);
      return true;
    } else {
      str += c;
      ++ptr;
    }
  }

  diag.print(diag::Diagnostic::error("Unclosed string")
                 .withLabel(diag::Label::primary(
                     offsetStart, (size_t)(ptr - start), "Unclosed string")));
  return false;
}

bool Lexer::lexIdentifier(Token &result, const char *identStart) {
  const char *cur = ptr;

  while (cur < end) {
    const char *next = cur;
    char32_t cp = decodeUTF8(next, end);
    if (!isIdentContinue(cp))
      break;
    cur = next;
  }

  std::string ident(identStart, cur - identStart);
  ptr = cur;

#define KW(name, str)                                                          \
  if (ident == str)                                                            \
    result.setKind(TokenKind::name);                                           \
  else
#include "belalang/Lexer/TokenKinds.def"
  {
    result.setKind(TokenKind::Ident);
  }

  result.setStr(ident);
  return true;
}

bool Lexer::lexNumber(Token &result, char first) {
  std::string num;
  num += first;
  bool hasDecimal = false;

  while (ptr < end) {
    if (isdigit(static_cast<unsigned char>(*ptr))) {
      num += *ptr;
      ++ptr;
    } else if (*ptr == '.' && !hasDecimal) {
      hasDecimal = true;
      num += '.';
      ++ptr;
    } else {
      break;
    }
  }

  result.setKind(hasDecimal ? TokenKind::LiteralFloat
                            : TokenKind::LiteralInteger);
  result.setStr(num);
  return true;
}

} // namespace lexer
} // namespace belalang
